lib_client/run_time.ml

```diff
@@ -0,0 +1,209 @@
+    else (* if us > 0 then *)
+      Fmt.pf ppf "%Ld.%03Ldus" us ns
+
+let cmp_floats v1 v2 = abs_float (v1 -. v2) < 0.0000001
+
+type timestamps =
+  | Queued of float
+  | Running of { queued_at : float; started_at : float }
+  | Finished of {
+      queued_at : float;
+      started_at : float option;
+      finished_at : float;
+    }
+[@@deriving show]
+
+type run_time_info =
+  | Cached
+    (* This indicates that the job never ran because cached results were used *)
+  | Queued_for of float (* elapsed time in seconds *)
+  | Running of { queued_for : float; ran_for : float }
+  | Finished of { queued_for : float; ran_for : float option }
+[@@deriving show]
+
+let queued_for = function
+  | Cached -> 0.
+  | Queued_for v -> v
+  | Running v -> v.queued_for
+  | Finished v -> v.queued_for
+
+let ran_for = function
+  | Cached -> 0.
+  | Queued_for _ -> 0.
+  | Running v -> v.ran_for
+  | Finished v -> Option.value ~default:0. v.ran_for
+
+let eq_timestamps (st1 : timestamps) (st2 : timestamps) =
+  match (st1, st2) with
+  | Queued v1, Queued v2 -> cmp_floats v1 v2
+  | Running v1, Running v2 ->
+      cmp_floats v1.queued_at v2.queued_at
+      && cmp_floats v1.started_at v2.started_at
+  | ( Finished
+        {
+          queued_at = queued_at1;
+          started_at = None;
+          finished_at = finished_at1;
+        },
+      Finished
+        {
+          queued_at = queued_at2;
+          started_at = None;
+          finished_at = finished_at2;
+        } ) ->
+      cmp_floats queued_at1 queued_at2 && cmp_floats finished_at1 finished_at2
+  | ( Finished
+        {
+          queued_at = queued_at1;
+          started_at = Some started_at1;
+          finished_at = finished_at1;
+        },
+      Finished
+        {
+          queued_at = queued_at2;
+          started_at = Some started_at2;
+          finished_at = finished_at2;
+        } ) ->
+      cmp_floats queued_at1 queued_at2
+      && cmp_floats started_at1 started_at2
+      && cmp_floats finished_at1 finished_at2
+  | Queued _, (Running _ | Finished _)
+  | Running _, (Queued _ | Finished _)
+  | ( Finished { started_at = None; _ },
+      (Queued _ | Running _ | Finished { started_at = Some _; _ }) )
+  | ( Finished { started_at = Some _; _ },
+      (Queued _ | Running _ | Finished { started_at = None; _ }) ) ->
+      false
+
+let run_times_from_timestamps ~build_created_at
+    ?(current_time = Unix.gettimeofday ()) (ts : timestamps) : run_time_info =
+  match ts with
+  | Queued v -> Queued_for (current_time -. v)
+  | Running { queued_at; started_at } ->
+      Running
+        {
+          queued_for = started_at -. queued_at;
+          ran_for = current_time -. started_at;
+        }
+  | Finished { queued_at; started_at; finished_at } -> (
+      if finished_at < build_created_at then Cached
+      else
+        match started_at with
+        | None ->
+            Finished { queued_for = finished_at -. queued_at; ran_for = None }
```

Comment:
"Finish" is an overloaded term here -- I think that it indicates "finished successfully" to us but in fact, is modelled simply as the point at which a job stops. 

It can never start -- for example it could be cancelled or aborted. It will still have a finished_at timestamp which would be the point at which it was cancelled or aborted.

So, no, it is not an error if a job is finished but never started. It just indicates that the job never ran. (Confusing I know :/ )